src/destructors.md

```diff
@@ -37,57 +33,344 @@ impl Drop for ShowOnDrop {
+* The parent of a statement scope is the scope of the block that contains the
+  statement.
+* The parent of the expression for a `match` guard is the scope of the arm that
+  it's for.
+* The parent of the expression for a given `match` arm is that arm's scope.
+* The parent of the arm scope is the scope of the match expression that it
+  belongs to.
+* The parent of all other scopes is the scope of the immediately enclosing
+  expression.
+
+### Scopes of function parameters
+
+All function parameters are in the scope of the entire function body, so are
+dropped last when evaluating the function. The actual function parameter is
+dropped after any named parameters that are bound to parts of it.
+
+```rust
+# struct ShowOnDrop(&'static str);
+# impl Drop for ShowOnDrop {
+#     fn drop(&mut self) {
+#         println!("drop({})", self.0);
+#     }
+# }
+// Drops the second parameter, then `y`, then the first parameter, then `x`
+fn patterns_in_parameters(
+    (x, _): (ShowOnDrop, ShowOnDrop),
+    (_, y): (ShowOnDrop, ShowOnDrop),
+) {}
+
+// drop order is 3 2 0 1
+patterns_in_parameters(
+    (ShowOnDrop("0"), ShowOnDrop("1")),
+    (ShowOnDrop("2"), ShowOnDrop("3")),
+);
+```
+
+### Scopes of local variables
+
+Local variables declared in a `let` statement are associated to the scope of
+the block that contains the `let` statement. Local variables declared in a
+match are associated to the arm scope of the `match` arm that they are declared
+in.
+
+```rust
+# struct ShowOnDrop(&'static str);
+# impl Drop for ShowOnDrop {
+#     fn drop(&mut self) {
+#         println!("drop({})", self.0);
+#     }
+# }
+let declared_first = ShowOnDrop("Dropped last in outer scope");
 {
-    let mut overwritten = ShowOnDrop("Drops when overwritten");
-    overwritten = ShowOnDrop("drops when scope ends");
-}
-# println!("");
-{
-    let declared_first = ShowOnDrop("Dropped last");
-    let declared_last = ShowOnDrop("Dropped first");
+    let declared_in_block = ShowOnDrop("Dropped in inner scope");
 }
-# println!("");
-{
-    // Tuple elements drop in forwards order
-    let tuple = (ShowOnDrop("Tuple first"), ShowOnDrop("Tuple second"));
+let declared_last = ShowOnDrop("Dropped first in outer scope");
+```
+
+If multiple patterns are used in the same arm for a match, then an unspecified
+pattern will be used to determine the drop order.
+
+### Temporary scopes
+
+The *temporary scope* of an expression is the scope that is used for the
+temporary variable that holds the result of that expression when used in a
+[place context], unless it is promoted to a `static`.
+
+Apart from lifetime extension, the temporary scope of an expression is the
+smallest scope that contains the expression and is for one of the following:
+
+* The entire function body.
+* A statement.
+* The body of a [`if`], [`while`] or [`loop`] expression.
+* The `else` block of an `if` expression.
+* The condition expression of an `if` or `while` expression, or a `match` guard.
```

Comment:
What about the scrutinee of `match` / `if let` / `while let` ?

Reply:
They're not a temporary scopes. For example in `match 1 { ref mut z => (), };` the temporary for `1` lives until the end of the statement

Reply:
That seems like noteworthy in the reference.